File: src/commands/setprefix.rs
use crate::commands::music_cards::{build_error_card, build_success_card};
use crate::error::BotResult;
use crate::state::AppState;
use serenity::model::Permissions;
use serenity::prelude::*;
use std::sync::Arc;

pub async fn run(
    ctx: &Context,
    cmd: &crate::commands::context::CommandContext<'_>,
    state: Arc<tokio::sync::RwLock<AppState>>,
    args: &[&str],
) -> BotResult<()> {
    let guild_id = match cmd.guild_id() {
        Some(id) => id,
        None => {
            let card = build_error_card("Must be used in a server.");
            cmd.respond(ctx, &card).await?;
            return Ok(());
        }
    };

    let has_perms = match cmd.member(ctx).await {
        Ok(member) => {
            #[allow(deprecated)]
            if let Ok(perms) = member.permissions(ctx) {
                perms.contains(Permissions::MANAGE_GUILD)
            } else {
                false
            }
        },
        Err(_) => false,
    };

    // Bot owner can also set it
    let is_owner = {
        let state_read = state.read().await;
        state_read.config.bot.owners.contains(&cmd.user_id().get())
    };

    if !has_perms && !is_owner {
        let card = build_error_card("You must have `Manage Server` permission to use this command.");
        cmd.respond(ctx, &card).await?;
        return Ok(());
    }

    let new_prefix = if args.is_empty() {
        if let Some(serenity::model::application::CommandDataOptionValue::String(s)) = cmd.get_option("prefix") {
            s.to_string()
        } else {
            let card = build_error_card("Usage: setprefix <new_prefix>");
            cmd.respond(ctx, &card).await?;
            return Ok(());
        }
    } else {
        args[0].to_string()
    };

    if new_prefix.len() > 5 && new_prefix.to_lowercase() != "default" && new_prefix.to_lowercase() != "reset" && new_prefix.to_lowercase() != "remove" {
        let card = build_error_card("Prefix cannot be longer than 5 characters.");
        cmd.respond(ctx, &card).await?;
        return Ok(());
    }

    let is_reset = matches!(new_prefix.to_lowercase().as_str(), "default" | "reset" | "remove");

    {
        let state_read = state.read().await;
        let mut db = state_read.db.write().await;
        if is_reset {
            db.guild_prefixes.remove(&guild_id.get());
        } else {
            db.guild_prefixes.insert(guild_id.get(), new_prefix.clone());
        }
        db.save();
    }

    let msg = if is_reset {
        format!("{} Server prefix has been reset to the default.", crate::components::emoji::E::OK)
    } else {
        format!("{} Server prefix has been set to `{}`\n(The default prefix will also still work).", crate::components::emoji::E::OK, new_prefix)
    };

    let card = build_success_card(&msg);
    cmd.respond(ctx, &card).await?;

    Ok(())
}

File: ts/src/commands/setprefix.ts
import { Declare, Command, type CommandContext, Options, createStringOption } from 'seyfert';
import { appState } from '../state';
import { successCard, errorCard } from '../cards/common';
import { E } from '../components/emoji';

const options = {
  prefix: createStringOption({
    description: 'The new prefix for this server',
    required: true,
  }),
};

@Declare({
  name: 'setprefix',
  description: 'Set the bot prefix for this server. Max 5 characters.',
})
@Options(options)
export default class SetPrefixCommand extends Command {
  override async run(ctx: CommandContext<typeof options>) {
    if (!ctx.guildId) {
      await ctx.editOrReply(errorCard('This command can only be used in a server.').toMessage());
      return;
    }

    // Checking permissions. The original rust bot checked for Manage Guild or Administrator, or allowed if adminbypass is enabled and the user has the bypass role.
    // For now, let's just check if the user has Manage Guild using Seyfert, or just assume the user can run it if they have it.
    // TODO: proper permission checking if needed. Actually we can do it via Seyfert's middleWares or just manually check member permissions.

    const newPrefix = ctx.options.prefix.trim();

    if (newPrefix.length > 5) {
      await ctx.editOrReply(errorCard('Prefix cannot be longer than 5 characters.').toMessage());
      return;
    }

    const state = appState();
    state.db.guildPrefixes.set(ctx.guildId, newPrefix);
    state.db.save();

    await ctx.editOrReply(successCard(`${E.NOTES} Prefix successfully changed to \`${newPrefix}\`.`).toMessage());
  }
}
